### PR TITLE
[WIP] Log Directories Created After Startup not Tracked

### PR DESCRIPTION
**Pull request by stsantilena**

### CODE REVIEW

File: common/fs/src/cache/dir_path.rs [hunk 1]
```diff
@@ -74,3 +75,59 @@ impl From<DirPathBuf> for PathBuf {
...
 }
+
+// Recursive directory back off
+fn find_valid_path(
+    path: Option<PathBuf>,
+    postfix: Option<PathBuf>,
+) -> Result<DirPathBuf, DirPathBufError> {
+    match path {
+        Some(p) => {
+            if p.is_dir() {
```
Comment (dkhokhlov) [inline common/fs/src/cache/dir_path.rs:86]: does it work for:
- it is a file
- permission error

File: common/fs/src/cache/dir_path.rs [hunk 2]
```diff
@@ -74,3 +75,59 @@ impl From<DirPathBuf> for PathBuf {
...
+                Ok(DirPathBuf { inner: p, postfix })
+            } else {
+                warn!("{:?} is not a directory; moving to parent directory", p);
+                let mut postfix_pathbuf = PathBuf::new();
+                let mut pop_pathbuf = PathBuf::new();
+                if let Some(path) = postfix {
```
Comment (dkhokhlov) [inline common/fs/src/cache/dir_path.rs:92]: shadowing fn arg on purpose?

File: common/fs/src/cache/dir_path.rs [hunk 3]
```diff
@@ -74,3 +75,59 @@ impl From<DirPathBuf> for PathBuf {
...
+        None => None,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_level_up() {
```
Comment (dkhokhlov) [inline common/fs/src/cache/dir_path.rs:123]: find_valid_path has various corner cases...

File: common/fs/src/cache/dir_path.rs (hunk 2 above)
Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs:92]: The `postfix` function input is only non-None in the event that a given directory doesn't exist, in which case it saves that piece of the path so it can properly construct the missing directory path.

File: common/fs/src/cache/dir_path.rs (hunk 3 above)
Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs:123]: True. In this case the `level_up` function was easy to cover in a simple unit test. There are other unit and integration test that do seem to cover path creation. Are there any specific cases you suggest testing?

File: common/fs/src/cache/dir_path.rs (hunk 1 above)
Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs:86]: The configuration is meant to be set as directory so a single file wouldn't be expected. I followed same pattern as existing directory check. As for permissions, in theory, I think could fail is for some reason it was running in non-root mode and didn't have directory access.

File: common/config/src/lib.rs [hunk 4]
```diff
@@ -344,6 +344,7 @@ impl TryFrom<RawConfig> for Config {
                 // Filter off paths that are not directories and warn about them
                 .filter_map(|d| {
                     d.clone()
+                        //TODO: Need to fix this iteration as warning now happens in dir_path.rs
```
Comment (stsantilena) [inline common/config/src/lib.rs:347]: @c-nixon I don't think there needs to any change here, perhaps the warning might be different message?

File: common/fs/src/cache/dir_path.rs [hunk 5]
```diff
@@ -27,18 +28,18 @@ impl Deref for DirPathBuf {
...
+            Ok(p) => Ok(p),
+            _ => Err(DirPathBufError::NotADirPath(path)),
         }
 
         #[cfg(windows)]
         {
-            Ok(DirPathBuf { inner: path })
+            Ok(DirPathBuf {
+                inner: path,
+                postfix: None,
```
Comment (stsantilena) [inline common/fs/src/cache/dir_path.rs:41]: Had to add this line to fix the Windows build. This shouldn't affect existing Windows behavior.

File: bin/tests/it/cli.rs [hunk 6]
```diff
@@ -1573,7 +1573,6 @@ async fn test_line_redact() {
     test_line_rules(None, None, redact, to_write, expected).await;
 }
 
-#[ignore]
 #[tokio::test]
 async fn test_directory_created_after_initialization() {
```
Comment (stsantilena) [inline bin/tests/it/cli.rs:1577]: Turned on bug specific integration test from previous PR.

File: common/fs/src/cache/dir_path.rs (hunk 3 above)
Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs:129]: @dkhokhlov there is also an integration test that was merged before this PR to specifically test, I activated it in the last commit.

File: common/fs/src/cache/dir_path.rs (hunk 2 above)
Comment (dkhokhlov) [reply common/fs/src/cache/dir_path.rs:98]: i was thinking about var name

File: common/fs/src/cache/dir_path.rs (hunk 3 above)
Comment (dkhokhlov) [reply common/fs/src/cache/dir_path.rs:129]: root, w/o any "/", file, invalid path

Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs:143]: Added additional code coverage around `find_valid_path` function.

File: common/fs/src/cache/dir_path.rs [hunk 7]
```diff
@@ -74,3 +83,111 @@ impl From<DirPathBuf> for PathBuf {
...
+                let root_pathbuf = Path::new("/");
+                let mut postfix_pathbuf = PathBuf::new();
+                let mut pop_pathbuf = PathBuf::new();
+                if let Some(path) = postfix {
+                    pop_pathbuf.push(path);
+                }
+                let parent = match level_up(&p) {
+                    Some(p) => {
+                        if p == root_pathbuf {
+                            panic!("configured directory recursed to root!");
```
Comment (stsantilena) [inline common/fs/src/cache/dir_path.rs:107]: Set to panic if function recurses back to the root directory. This will watch the entire root file system which is most likely not desirable.

File: common/fs/src/cache/dir_path.rs (hunk 2 above)
Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs]: Ah, I see. Updated var name `path` -> `pop`

File: common/fs/src/cache/dir_path.rs (hunk 7 above)
Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs]: Changed this so that instead of throwing a panic, it will warn and watch root level for missing dir.

File: common/config/src/lib.rs (hunk 4 above)
Comment (dkhokhlov) [reply common/config/src/lib.rs:347]: if we allow to specify non-existing dirs - why do we filter them here?

Comment (stsantilena) [reply common/config/src/lib.rs]: This iteration just filters out anything that throws an error. Originally, if a directory didn't exist and error was thrown. Now, there is a more complex process happening down in `find_valid_path` that allows directories that don't yet exist; still propagating errors up but for different reasons. So really just the comment is no longer valid.

**Review by c-nixon — APPROVED** [review]
This all looks good to me!

File: common/fs/src/cache/dir_path.rs [hunk 8]
```diff
@@ -74,3 +83,183 @@ impl From<DirPathBuf> for PathBuf {
...
+    postfix: Option<PathBuf>,
+) -> Result<DirPathBuf, DirPathBufError> {
+    match path {
+        Some(p) => {
+            if matches!(std::fs::canonicalize(&p), Ok(_)) {
+                Ok(DirPathBuf { inner: p, postfix })
+            } else {
+                warn!("{:?} is not a directory; moving to parent directory", p);
+                let root_pathbuf = Path::new("/");
+                let mut postfix_pathbuf = PathBuf::new();
```
Comment (c-nixon) [inline common/fs/src/cache/dir_path.rs:103]: I feel like there might be some way to use the std::path `Component` types to implement this, but that might just complicate the logic

Comment (stsantilena) [reply common/fs/src/cache/dir_path.rs:103]: There's probably a few different ways to go about it. The tricky part is flipping the path each integration to make sure it comes out correct, and plays nicely with the path rules.
